Include/cpython/object.h

```diff
@@ -418,83 +418,33 @@ PyAPI_FUNC(void) _Py_NO_RETURN _PyObject_AssertFailed(
-above 0 forever, and the trashcan will never get emptied.
-
-How it works:  The BEGIN macro increments a call-depth counter.  So long
-as this counter is small, the body of the deallocator is run directly without
-further ado.  But if the counter gets large, it instead adds p to a list of
-objects to be deallocated later, skips the body of the deallocator, and
-resumes execution after the END macro.  The tp_dealloc routine then returns
-without deallocating anything (and so unbounded call-stack depth is avoided).
-
-When the call stack finishes unwinding again, code generated by the END macro
-notices this, and calls another routine to deallocate all the objects that
-may have been added to the list of deferred deallocations.  In effect, a
-chain of N deallocations is broken into (N-1)/(Py_TRASHCAN_HEADROOM-1) pieces,
-with the call stack never exceeding a depth of Py_TRASHCAN_HEADROOM.
-
-Since the tp_dealloc of a subclass typically calls the tp_dealloc of the base
-class, we need to ensure that the trashcan is only triggered on the tp_dealloc
-of the actual class being deallocated. Otherwise we might end up with a
-partially-deallocated object. To check this, the tp_dealloc function must be
-passed as second argument to Py_TRASHCAN_BEGIN().
-*/
-
-/* Python 3.9 private API, invoked by the macros below. */
+/* Python 3.9 private API */
 PyAPI_FUNC(int) _PyTrash_begin(PyThreadState *tstate, PyObject *op);
 PyAPI_FUNC(void) _PyTrash_end(PyThreadState *tstate);
 
 PyAPI_FUNC(void) _PyTrash_thread_deposit_object(PyThreadState *tstate, PyObject *op);
 PyAPI_FUNC(void) _PyTrash_thread_destroy_chain(PyThreadState *tstate);
 
 
-/* Python 3.10 private API, invoked by the Py_TRASHCAN_BEGIN(). */
+/* Python 3.10 private API */
 
 /* To avoid raising recursion errors during dealloc trigger trashcan before we reach
  * recursion limit. To avoid trashing, we don't attempt to empty the trashcan until
- * we have headroom above the trigger limit */
+ * we have headroom above the trigger limit
+ *
+ * FIXME: could this be moved into object.c?
+ */
 #define Py_TRASHCAN_HEADROOM 50
 
+/* Backwards compatibility macros.  The trashcan mechanism is now integrated
+ * with _Py_Dealloc and so these do nothing.  The assert() is present to handle
+ * the case that a 'goto' statement precedes Py_TRASHCAN_END.  */
 #define Py_TRASHCAN_BEGIN(op, dealloc) \
-do { \
-    PyThreadState *tstate = PyThreadState_Get(); \
-    if (tstate->c_recursion_remaining <= Py_TRASHCAN_HEADROOM && Py_TYPE(op)->tp_dealloc == (destructor)dealloc) { \
-        _PyTrash_thread_deposit_object(tstate, (PyObject *)op); \
-        break; \
-    } \
-    tstate->c_recursion_remaining--;
-    /* The body of the deallocator is here. */
+    do {
+
 #define Py_TRASHCAN_END \
-    tstate->c_recursion_remaining++; \
-    if (tstate->delete_later && tstate->c_recursion_remaining > (Py_TRASHCAN_HEADROOM*2)) { \
-        _PyTrash_thread_destroy_chain(tstate); \
-    } \
-} while (0);
+    assert(1); \
```

Comment:
What is this assert(1)?

Reply:
If you look at the resulting code and not the diff, there is a comment right above it:

> The assert() is present to handle the case that a 'goto' statement precedes Py_TRASHCAN_END.

I think maybe only some compilers complain about it but `assert(1)` seemed like an easy fix.